Acids, Smoke, Foam, reagents reaction() refactors

phil235 opened the pull request:
- Reagents.reaction() proc and subprocs:
  - fixed the fraction argument in reaction(...., INGESTED, fraction) so the proper amounts of each reagent react, not all of it. (e.g. eating a snack called reaction(..., INGESTED, 1), basically reacting with all reagents instead of only the ones in the bite we take)
  - some optimizations.
  - refactor how protection against touched chems being absorbed is done ( /datum/reagent/proc/reaction_mob(...,TOUCH) was fucked, you could be naked with only a pair of medical shoes and you wouldn't absorb any reagent at all). I created the proc get_permeability_protection() to get that protection. No more simple RNG get touched or not, now the amount that touches you depends on your protection and the volume in question. You now need a full bio suit or equivalent, plus a gas mask or internals on, to be fully protected against the default touch reactions that transfer reagents to your body (but if it's a reagent with its own reaction_mob() proc like acid, you're still fucked). Monkey's mask no longer give it them full protection. The xeno's hard skin give it a decent natural protection. 
  - Adds an argument to the add_reagent() proc to block automatic call of handle_reaction()
  - When using reagents/proc/trans_to(), reactions are now only handled AFTER every reagent is transfered and not before. (fixes foam cleaner grenade containing no space cleaner)
- Smoke/Foam effect changes:
  - All smoke effects now uses process() just like foam.
  - The number of smoke clouds depends on the amount of smoke reagent created.
  - The amount of reagents distributed by the smoke is now never more than the total amount, it is also evenly distributed. Fixes #8060. Fixes #8498. Takes care of half of #8647.
  - fixes not ingesting smoke if it's in the dark.
  - removed BLOCK_GAS_SMOKE_EFFECT flag from clown mask and other non "actual" gas mask.
  - fixes metalfoam not working. Fixes #9216 
  - effect/smoke is made more OOP 
  - Fixes many reagents having effects on mobs not dependent on volume. Fixes #8316. Fixes #9868. 
    Fixes #10107 
  - Remove show_message argument from blob reagent reaction_mob()'s and remove the if(show_message) send_message(). Now send_message is called directly on each blob strike.
- ACID:
  - Damage on humans now depends on acidpwr&volume.
  - changes item/acid_act to take volume into account.  Add a volume check to item/acid_act(), there's a threshold to be able to melt and then a probability to melt (acidpwr&volume). Fixes #6857.
  - When calling human/acid_act(), item acid melting chance lowers after each  items it tries to melt in the list, some of the acid is "used" to melt that item so the next items have a lower chance to melt (to make melting every clothing harder).
  - Sprays can now contain acid again.. Fixes #6276. Fixes #8179.
  - melted storage items now drop their content instead of deleting everything.
- Chem grenade:
  - chem_grenade reaction, if there's no reagent left after reaction (smoke/foam/etc, reagents cleared) you don't get the steam effect and immediate reaction with all atoms around the grenade explosion. (Fixes foam cleaner grenade deleting bloodstains around it as soon as the grenade explodes, even if the foam hasn't reached the tile yet).
  - removing some useless code in chem_grenade/prime().

phil235 committed:
Lots of work on acids, smoke, foam, reagents reaction() proc.

- fixed the fraction argument in reaction(...., INGESTED, fraction) so the proper amounts of each reagent react, not all of it.
- slight optimization of datum/reagents/reaction()
- small optimization /datum/reagent/proc/reaction_mob
- the smoke's reagents no longer do direct transfer upon crossed()/move() but uses "reagents.reaction( mob, TOUCH)" only.
- fixes no ingesting smoke if it's in the dark.
- removed block gas smoke effect flag from clown mask and other non "actual" gas mask.
- refactor how protection against touched chems being absorbed is done. get_permeability_protection() to get that protection. No more

simple RNG get touched or not, now the amount that touches you depends on your protection and the volume in question.
- changes acid_act to take volume into account.
- chem_smoke now uses process() just like foam, we remove the reaction from crossed() and Move(), it's only done in process() now.
- optimization of effect/smoke, OOP! And now all smoke effects use objprocessing instead of shitty sleep/spawn
- removed mob var/coughedtime.
- add a volume check to item/acid_act(), there's a threshold to be able to melt and then a probability to melt (acidpwr&volume).
- greatly lowered the amount of reagent in the smoke from dead blobspore (so the volume isn't enough for melting)
- human/acid_act(): damage to limbs depends on acidpwr&volume.
- fixes metalfoam not working.
- smoke powder: only one start() call.
- Amound of smoke objects depends on amount of smoke recipe created.
- Adds an argument to add_reagent() to block automatic call of handle_reaction()
- When using reagents/proc/trans_to(), reactions are now only handled after every reagent is transfered and not before.
- Amount of smoke objects depends on amount of smoke reagent created.
- radius of foam reaction depends on amount of foam reagent created.
- The amount of other reagents inside the smoke/foam decides the life expectancy of the effect.
- The amount of reagents in each small smoke cloud/foam cell depends on the amount of other reagents in the initial reaction but also

how much smoke/foam was created (more smoke means dilution of the reagents).
- smoke/foam's reagent reaction on mob decreases the life expectancy of the effect (to avoid reagent duplication)
- The amount of reagent reacting with atoms is less if the life expectancy is high.(to avoid reagent duplication)
- The amount of movement from the smoke now depends on the number of smoke clouds created.
- removing some useless code in chem_grenade/prime().
- When calling human/acid_act(), item acid melting chance lowers after each each successful melt in the list, some of the acid is

"used" to melt that item so the next items have a lower chance to melt (to make melting every clothing harder).
- remove the banned reagent list from sprays, not needed anymore now that acid is nerfed.
- chem_grenade reaction, if there's no reagent left after reaction (smoke/foam/etc, reagents cleared) you don't get the steam effect

and immediate reaction with all atoms around the grenade explosion. (Fixes foam cleaner grenade deleting bloodstains around it as soon

as the grenade explodes, even if the foam hasn't reached the tile yet).
- melted storage items now drop their content instead of deleting everything.

Iamgoofball commented:
:+1:

WJohn commented:
Jesus christ.

Lumipharon commented:
Could you give a few examples to show how these changes will behave in practice?
IE: how much smoke reagent you need for X radius
How long it last/how fast it transfers etc.

>The amount of other reagents inside the smoke/foam decides the life expectancy of the effect.
Does that mean if you have just smoke, it's super short lived?

phil235 committed:
a few fixes

phil235 commented:
>Does that mean if you have just smoke, it's super short lived?
There's a minimum lifetime.

Alright, let's get some examples:
foam radius is minimum 1 (a "plus") and can go up to 20.
Amount of smoke clouds is 1-20.

Razharas commented:
If only reaction of touch on mob would get a target argument, we could have made real good patches

phil235 committed:
another fix and smoother smoke fading.

freerealestate commented:
That sure is a lot of bullet points I won't read

On 22 June 2015 at 11:55, Razharas <notifications@github.com> wrote:

> If only reaction of touch on mob would get a target argument, we could
> have made real good patches
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tgstation/-tg-station/pull/10085#issuecomment-113970031>
> .
>

KorPhaeron commented:
>removed BLOCK_GAS_SMOKE_EFFECT flag from clown mask and other non "actual" gas mask.

This whole PR is a smokescreen for a stealth clown nerf

Iamgoofball commented:
@razharas that idea is bad and we've gone over why it's bad
On Jun 21, 2015 4:55 PM, "Razharas" <notifications@github.com> wrote:

> If only reaction of touch on mob would get a target argument, we could
> have made real good patches
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tgstation/-tg-station/pull/10085#issuecomment-113970031>
> .
>

Jalleo commented on code/modules/clothing/masks/gasmask.dm:
```diff
@@ -222,7 +222,7 @@
 /obj/item/clothing/mask/gas/clown_hat
 	name = "clown wig and mask"
 	desc = "A true prankster's facial attire. A clown is incomplete without his wig and mask."
-	flags = MASKCOVERSEYES | BLOCK_GAS_SMOKE_EFFECT | MASKINTERNALS
```
RIP clown protection against smoke.

phil235 commented:
I still need to do more tests to adjust the different values, maybe tweak the reagents in the hard coded chem grenades, plus add a changelog.

phil235 committed:
- Foam/moke lifetime is no longer affected by reagents volume.
- Remove show_message argument from blob reagent reaction_mob()'s and remove the if(show_message) send_message(). Now send_message is called directly on each blob strike.
- Changed blob reagent effects to depend on volume so blob spore smoke isn't deadly.
- Breathing smoke now reduces its life expectancy and calls reaction(INGEST) with the accurate amount of reagent.
- Adjusted the effects on mobs of some other reagents.
- Adjusted reagent amounts in chem grenades.

phil235 committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into AcidFix

Conflicts:
	code/modules/reagents/Chemistry-Reagents/Blob-Reagents.dm

optimumtact commented:
@phil235 would love to see this soon

phil235 committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into AcidFix

Conflicts:
	code/modules/clothing/masks/gasmask.dm

phil235 committed:
Derp, I hate fixing merge conflicts...

phil235 commented:
Alright, should be ready now.
